manager/state/raft/raft.go

```diff
@@ -213,32 +214,45 @@ func NewNode(ctx context.Context, opts NewNodeOptions) (*Node, error) {
 	n.reqIDGen = idutil.NewGenerator(uint16(n.Config.ID), time.Now())
 	n.wait = newWait()
 
-	if n.startNodePeers != nil {
+	if loadAndStartErr == errNoWAL {
 		if n.joinAddr != "" {
 			c, err := n.ConnectToMember(n.joinAddr, 10*time.Second)
 			if err != nil {
 				return nil, err
 			}
+			client := api.NewRaftMembershipClient(c.Conn)
 			defer func() {
 				_ = c.Conn.Close()
 			}()
 
 			ctx, cancel := context.WithTimeout(n.Ctx, 10*time.Second)
 			defer cancel()
-			resp, err := c.Join(ctx, &api.JoinRequest{
+			resp, err := client.Join(ctx, &api.JoinRequest{
 				Addr: n.Address,
 			})
 			if err != nil {
 				return nil, err
 			}
 
+			n.Config.ID = resp.RaftID
+
+			if _, err := n.createWAL(opts.ID); err != nil {
+				return nil, err
+			}
+
 			n.Node = raft.StartNode(n.Config, []raft.Peer{})
 
 			if err := n.registerNodes(resp.Members); err != nil {
 				return nil, err
 			}
 		} else {
-			n.Node = raft.StartNode(n.Config, n.startNodePeers)
+			// First member in the cluster, self-assign ID
+			n.Config.ID = uint64(rand.Int63()) + 1
```

Comment:
I don't think there's any need for this to be a secure random. We always test for collisions and retry. We could just as well increment a counter instead of randomly generating the IDs.